datasetinsights/stats/visualization/coco_pose.py

```diff
@@ -0,0 +1,305 @@
+        kp_dict[name] = {"x": [], "y": []}
+
+    for ann in annotations:
+        keypoints = ann["keypoints"]
+        keypoints = _remove_visibility_flag_from_kp(keypoints)
+
+        # Separate x and y keypoints
+        x_kp, y_kp = keypoints[0::2], keypoints[1::2]
+        x_kp, y_kp = _translate_and_scale_xy_kp(x_kp, y_kp)
+
+        # save keypoints to dict
+        idx = 0
+        for xi, yi in zip(x_kp, y_kp):
+            if xi == 0 and yi == 0:
+                pass
+            else:
+                kp_dict[COCO_KEYPOINTS[idx]]["x"].append(xi)
+                kp_dict[COCO_KEYPOINTS[idx]]["y"].append(yi)
+            idx += 1
+
+    return kp_dict
+
+
+def generate_scatter_plot(kp_dict, title="", fig_path=None):
+    fig, ax = plt.subplots(dpi=300, figsize=(8, 8))
+    colors = plt.cm.rainbow(np.linspace(0, 1, len(COCO_KEYPOINTS)))[::-1]
+    i = 0
+    for name in COCO_KEYPOINTS:
+        plt.scatter(
+            kp_dict[name]["x"],
+            kp_dict[name]["y"],
+            color=colors[i],
+            label=name,
+            s=3,
+        )
+        i += 1
+
+    plt.xlim(-2, 2)
+    plt.ylim(-2, 2)
+
+    ax.spines["left"].set_position("center")
+    ax.spines["bottom"].set_position("center")
+
+    # Eliminate upper and right axes
+    ax.spines["right"].set_color("none")
+    ax.spines["top"].set_color("none")
+
+    # Turn off tick labels
+    ax.set_yticklabels([])
+    ax.set_xticklabels([])
+    ax.set_title(title)
+
+    # Invert axes
+    ax.invert_yaxis()
+    ax.set_aspect("equal")
+
+    plt.legend(
+        loc="lower center", bbox_to_anchor=(0.5, -0.17), ncol=5, fancybox=True,
+    )
+
+    if fig_path:
+        plt.savefig(fig_path, bbox_inches="tight", pad_inches=0.15)
+    plt.show()
+
+
+def generate_heatmaps(kp_dict, color="red", title="", fig_dest=None):
+
+    for name in COCO_KEYPOINTS:
+        fig, ax = plt.subplots(dpi=100, figsize=(8, 8))
+        sns.kdeplot(
+            x=kp_dict[name]["x"],
+            y=kp_dict[name]["y"],
+            shade=True,
+            label=name,
+            ax=ax,
+            alpha=0.85,
+            color=color,
+            cbar=False,
+        )
+
+        ax.set_xlim(-2.5, 2.5)
+        ax.set_ylim(-2.5, 2.5)
+
+        ax.spines["left"].set_position("center")
+        ax.spines["bottom"].set_position("center")
+
+        # Eliminate upper and right axes
+        ax.spines["right"].set_color("none")
+        ax.spines["top"].set_color("none")
+
+        # Turn off tick labels
+        ax.set_yticklabels([])
+        ax.set_xticklabels([])
+
+        t = name.split("_")
+        t = [x.capitalize() for x in t]
+        t = " ".join(t)
+        textstr = t
+        ax.text(
+            0.005, 0.080, textstr, transform=ax.transAxes, fontsize=36, va="top"
+        )
+
+        ax.patch.set_edgecolor("black")
+
+        ax.patch.set_linewidth("1")
+
+        # Invert axes
+        ax.invert_yaxis()
+
+        ax.set_aspect("equal")
+
+        if fig_dest:
```

Comment:
Let's decouple saving to file from the "generate plot" itself. The plot generate method should not have any side effects like this. Otherwise, it going to be very hard to write unittests.